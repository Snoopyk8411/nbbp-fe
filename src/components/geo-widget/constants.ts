import { IGeoPoint } from 'tools/types/geolocation-types';

export enum YMAP_CONFIG {
  LANG = 'ru_RU',
  PACKAGE = 'package.full',
  CLUSTERER_PRESET = 'invertedVioletClusterIcons',
  MODULE_BALLOON = 'geoObject.addon.balloon',
  MODULE_HINT = 'geoObject.addon.hint',
}

export const GEO_POINTS: IGeoPoint[] = [
  {
    id: '1',
    title: 'Москва 1',
    tag: 'МСК1',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Nostrum omnis obcaecati natus, ea repellendus necessitatibus aut assumenda blanditiis commodi, officia fugiat similique repellat ratione at. Sunt, sequi illum? Dicta, dolorem!',
    coords: [55.831903, 37.411961],
  },
  {
    id: '2',
    title: 'Москва 2',
    tag: 'МСК2',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Nostrum omnis obcaecati natus, ea repellendus necessitatibus aut assumenda blanditiis commodi, officia fugiat similique repellat ratione at.',
    coords: [55.763338, 37.565466],
  },
  {
    id: '3',
    title: 'Москва 3',
    tag: 'МСК3',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Nostrum omnis obcaecati natus, ea repellendus necessitatibus aut assumenda blanditiis commodi, officia fugiat similique repellat ratione at.',
    coords: [55.769338, 37.595466],
  },
  {
    id: '4',
    title: 'Москва 4',
    tag: 'МСК4',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Nostrum omnis obcaecati natus, ea repellendus necessitatibus aut assumenda blanditiis commodi, officia fugiat similique repellat ratione at.',
    coords: [55.744522, 37.616378],
  },
  {
    id: '5',
    title: 'Череповец',
    tag: 'ЧЕР',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Nostrum omnis obcaecati natus, ea repellendus necessitatibus aut assumenda blanditiis commodi, officia fugiat similique repellat ratione at.',
    coords: [59.1257843, 37.9827323],
  },
];

export const INITIAL_MAP_STATE = {
  center: [55.751574, 37.573856],
  zoom: 5,
};

export const NO_LOCATION_TAG = 'N/A';
