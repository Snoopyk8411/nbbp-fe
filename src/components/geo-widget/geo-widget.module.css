.geo {
  width: 5rem;
  height: 5rem;
  position: relative;
}

.geo__toggle {
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: space-around;
  align-items: center;
  flex-direction: column;
  border-radius: 1.25rem;
  padding: 0.5rem 1rem;
  user-select: none;
  cursor: pointer;
  outline: none;
  border: none;
  color: var(--clr-neutral-400);
  background-color: var(--clr-neutral-100);
  position: relative;
}

.geo__toggle:hover {
  background-color: var(--clr-neutral-300);
}

.geo__value {
  font-size: 1.25rem;
  line-height: 1.2;
  color: var(--clr-neutral-400);
}

.geo__dropdown {
  position: absolute;
  background-color: var(--clr-neutral-100);
  width: min(100vw, 60rem);
  aspect-ratio: 1;
  display: none;
}

.geo.open .geo__dropdown {
  display: block;
}
