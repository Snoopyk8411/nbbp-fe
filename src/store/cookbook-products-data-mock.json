[
  {
    "id": 1,
    "name": "tomatoes"
  },
  {
    "id": 2,
    "name": "potatoes"
  },
  {
    "id": 3,
    "name": "olive oil"
  },
  {
    "id": 4,
    "name": "shrimps"
  },
  {
    "id": 5,
    "name": "beef"
  },
  {
    "id": 6,
    "name": "salad leafs"
  },
  {
    "id": 7,
    "name": "spaghetti"
  },
  {
    "id": 8,
    "name": "soar creame"
  },
  {
    "id": 9,
    "name": "chicken fillet"
  },
  {
    "id": 10,
    "name": "oyster"
  }
]
