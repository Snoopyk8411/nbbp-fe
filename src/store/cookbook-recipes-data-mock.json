[
  {
    "id": 0,
    "name": "Salad",
    "ingredients": [
      {
        "product": {
          "id": 1,
          "name": "tomato"
        },
        "amount": 10
      },
      {
        "product": {
          "id": 6,
          "name": "salad leafs"
        },
        "amount": 10
      },
      {
        "product": {
          "id": 3,
          "name": "olive oil"
        },
        "amount": 10
      }
    ]
  }
]
