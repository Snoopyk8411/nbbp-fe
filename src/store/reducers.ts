import { combineReducers } from '@reduxjs/toolkit';
import { productsSlice } from './contributors/Ilia_Kotov/cookbookProducts/slice';
import { recipesSlice } from './contributors/Ilia_Kotov/cookbookRecipes/slice';

import { galleryPageSlice } from './contributors/ivanefimov/slice';
import { welcomePageSlice } from './welcome/slice';
import { mediaSlice } from './gleb/slice';
import { todosPageSlice } from './aleksei/todos/slice';
import { shopSlice } from './shop/slice';

const rootReducer = combineReducers({
  welcomePage: welcomePageSlice.reducer,
  products: productsSlice.reducer,
  recipes: recipesSlice.reducer,
  todosPage: todosPageSlice.reducer,
  galleryPage: galleryPageSlice.reducer,
  mediaPage: mediaSlice.reducer,
  shop: shopSlice.reducer,
});

export default rootReducer;
export type RootState = ReturnType<typeof rootReducer>;
