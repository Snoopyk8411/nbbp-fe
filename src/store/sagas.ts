import { all } from 'redux-saga/effects';

import recipesLoadDataWatcher from './cookbookRecipes/saga-load-recipes';
import wpLoadInitDataWatcher from './welcome/saga-load-init-data';

export default function* indexSagas(): Generator<any> {
  yield all([wpLoadInitDataWatcher(), recipesLoadDataWatcher()]);
}
