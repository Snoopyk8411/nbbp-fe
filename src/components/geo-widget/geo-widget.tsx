import ReactDOM from 'react-dom';
import cn from 'classnames';
import { useCallback, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { renderToString } from 'react-dom/server';
import { YMaps, Map, Clusterer, Placemark, YMapsApi } from 'react-yandex-maps';

import {
  getCoords,
  getNearest,
  readCoordsFromLocalStorage,
  saveCoordsToLocalStorage,
} from 'tools/geolocation/geolocation';
import { GeoBalloon } from 'components/geo-balloon/geo-balloon';
import { ICoords } from 'tools/types/geolocation-types';
import GeoIcon from 'assets/geo.svg';
import { setGeo, setIsGeoOpen } from 'store/shop/slice';
import { selectGeo, selectIsGeoOpen } from 'store/shop/selectors';
import { GEO_POINTS, INITIAL_MAP_STATE, NO_LOCATION_TAG, YMAP_CONFIG } from './constants';
import geoStyles from './geo-widget.module.css';

export const GeoWidget = (): JSX.Element => {
  const [ymaps, setYmaps] = useState<YMapsApi>();
  const dropDownStyles = geoStyles.open || '';

  const dispatch = useDispatch();
  const location = useSelector(selectGeo);
  const isDropDownOpen = useSelector(selectIsGeoOpen);

  const updateLocation = useCallback((coords: ICoords): void => {
    const nearest = getNearest(coords, GEO_POINTS);
    if (nearest) {
      saveCoordsToLocalStorage(nearest.coords);
      dispatch(setGeo(nearest));
    }
    dispatch(setIsGeoOpen(false));
  }, []);

  useEffect(() => {
    const coords = readCoordsFromLocalStorage();
    if (coords) {
      updateLocation(coords);
    } else {
      const handleSuccess: PositionCallback = pos => {
        const coords = getCoords(pos);
        updateLocation(coords);
      };
      const handleError: PositionErrorCallback = err => {
        console.error(err.message);
      };
      navigator.geolocation.getCurrentPosition(handleSuccess, handleError);
    }
  }, []);

  const toggleDropdown = useCallback(() => {
    dispatch(setIsGeoOpen(true));
  }, []);

  return (
    <div className={cn(geoStyles.geo, { [dropDownStyles]: isDropDownOpen })}>
      <button className={geoStyles.geo__toggle} onClick={toggleDropdown} type='button'>
        <GeoIcon />
        <span className={geoStyles.geo__value}>{location?.tag || NO_LOCATION_TAG}</span>
      </button>
      <div className={geoStyles.geo__dropdown}>
        <YMaps query={{ lang: YMAP_CONFIG.LANG, load: YMAP_CONFIG.PACKAGE }}>
          <Map
            defaultState={INITIAL_MAP_STATE}
            width={'100%'}
            height={'100%'}
            onLoad={(ymaps): void => setYmaps(ymaps)}
          >
            <Clusterer
              options={{
                preset: YMAP_CONFIG.CLUSTERER_PRESET,
                groupByCoordinates: false,
                balloonPanelMaxMapArea: Infinity,
              }}
            >
              {ymaps &&
                GEO_POINTS.map(point => (
                  <Placemark
                    modules={[YMAP_CONFIG.MODULE_BALLOON, YMAP_CONFIG.MODULE_HINT]}
                    key={point.id}
                    geometry={point.coords}
                    onBalloonOpen={(): void => {
                      ReactDOM.hydrate(
                        <GeoBalloon point={point} handleSelect={(): void => updateLocation(point.coords)} />,
                        document.getElementById('balloon'),
                      );
                    }}
                    properties={{
                      hintContent: point.title,
                      balloonContent: `<div id="balloon">${renderToString(
                        <GeoBalloon point={point} handleSelect={(): void => updateLocation(point.coords)} />,
                      )}</div>`,
                    }}
                  />
                ))}
            </Clusterer>
          </Map>
        </YMaps>
      </div>
    </div>
  );
};
