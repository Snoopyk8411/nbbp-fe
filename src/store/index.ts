import createSagaMiddleware from 'redux-saga';

import { createStore, applyMiddleware, compose, Store } from '@reduxjs/toolkit';

import { rootReducer } from './reducers';
import sagas from './sagas';

const composeEnhancers =
  (typeof window !== 'undefined' && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

export const configureStore = (): Store => {
  const sagaMiddleware = createSagaMiddleware();
  const middlewares = [sagaMiddleware];
  const createdStore = createStore(rootReducer, composeEnhancers(applyMiddleware(...middlewares)));

  sagaMiddleware.run(sagas);

  return createdStore;
};

const store = configureStore();
export default store;
