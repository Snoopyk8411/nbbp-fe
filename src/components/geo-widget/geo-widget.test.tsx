import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { configureStore } from 'store';

import { GeoWidget } from './geo-widget';

describe('Geo component', () => {
  it('should get valid snapshot', () => {
    const mockStore = configureStore();
    const component = renderer.create(
      <Provider store={mockStore}>
        <GeoWidget />
      </Provider>,
    );
    expect(component).toMatchSnapshot();
  });
});
