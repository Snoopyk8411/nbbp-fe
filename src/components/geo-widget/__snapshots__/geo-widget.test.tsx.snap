// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Geo component should get valid snapshot 1`] = `
<div
  className="geo"
>
  <button
    className="geo__toggle"
    onClick={[Function]}
    type="button"
  >
    <test-file-stub />
    <span
      className="geo__value"
    >
      N/A
    </span>
  </button>
  <div
    className="geo__dropdown"
  />
</div>
`;
